[DesignPR] Controller Unit Testability Part 2

ianhong opened the pull request:
@rynowak  and @yishaigalatzer 

Could you take a look at this design PR? I'm also attaching the summary of this PR. 

Thanks

What are covered or tested in this Design PR
1. ViewBag
2. ModelState
3. Context.Request.GetFormAsync();formCollection.GetValues("PromoCode").FirstOrDefault()
7. Url.Action
5. Context.RequestAborted
9. Url.IsLocalUrl(returnUrl

Not yet implemented:
1. Context.User.Identity.GetUserName(), Context.User.Identity.GetUserId(), User.Identity.IsAuthenticated
2. Context.Request.Scheme
3. Response.SignOut
4. Need to expand the UrlAction callback signature
5. Context.Request.Cookies.Get("Session"); Context.Request.Cookies.Append("Session", cartId);
9. Context.Request.ReadFormAsync();formParameters["RequestVerificationToken"];

External Components to Test:
1. DbContext
2. IMemoryCache
3. IConnectionManager
4. UserManager<ApplicationUser>
5. SignInManager<ApplicationUser>
6. IHostingEnvironment
7. AntiForgery
8. IOptions
9. RoleManager<IdentityRole>
10. IFlashStrategy, ILogger

Exploratory Test
1 Adding controller unit tests for MusicStore
2. Updating unit tests for KExpense

rynowak commented on test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs:
```diff
@@ -0,0 +1,173 @@
+			public IActionResult UsingUrlAction(string action)
+			{
+				return View(Url.Action(action));
+			}
```
Would really like to see some more real-world examples here. These are a bit simple and are contrived towards the features that you targeted. Can we take some code from music store and mock out the db to use that as a starting point?

ianhong replied:
@rynowak In fact, my plan is adding controller unit tests for MusicStore once this feature is done. So, that's why these unit tests are simpler. I'm wondering if you still want to make these unit tests more realistic in addition to that??

yishaigalatzer replied:
Its probably better to build a test sample on top of music store and send as a separate PR (in the Music Store repo). I agree though the scenarios should be more realistic.

rynowak commented on src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs:
```diff
@@ -0,0 +1,277 @@
+			return controller;
+		}
+
+		private class MvcControllerTestHelperHttpContext : HttpContext
```
We should use `DefaultHttpContext` here. My guess is that we want any testing helper code to go into a separate project and not `Mvc.Core`. Then it's no trouble to introduce the dependency on the package that `DefaultHttpContext` lives in.

yishaigalatzer replied:
:+1:

rynowak commented on src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs:
```diff
@@ -0,0 +1,277 @@
+using Microsoft.Framework.DependencyInjection.Fallback;
+
+// Not sure if this is the right location/namespace for this..
+namespace Microsoft.AspNet.Mvc.Core
```
Discuss with @yishaigalatzer - my assumption is that we want another separate package - something like `Microsoft.AspNet.Mvc.Testing`. 

In general though there is no product code in the `Microsoft.AspNet.Mvc.Core` namespace. It should be `Microsoft.AspNet.Mvc` if it's something that a user will need to find easily, otherwise `Microsoft.AspNet.Mvc.<FeatureName>` if it's something obscure.

yishaigalatzer replied:
Yes this should be in another package

rynowak commented on src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs:
```diff
@@ -0,0 +1,277 @@
+// Not sure if this is the right location/namespace for this..
+namespace Microsoft.AspNet.Mvc.Core
+{
+	public class MvcControllerUnitTestHelper
```
What's the value of this class being instance-based? I don't see any of your examples leveraging that.

yishaigalatzer replied:
the name can be just `ControllerTestHelper`

ianhong replied:
Agree. It was static before, but while I'm trying to subclass UrlHelper, I made it a regular type and forgot to reverse. Thanks.

rynowak commented on src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs:
```diff
@@ -0,0 +1,277 @@
+				throw new NotImplementedException();
+			}
+
+			public bool IsLocalUrl(string url) // Can it be a static method or the UrlHelper class or UrlHelperExtensions??
```
You can subclass `UrlHelper` (with some tweaks) - or what's probably better is to move this to a static method and have `UrlHelper` and the testable version both call that method.

ianhong replied:
I think the latter (static method) should be better. I tried to subclass UrlHelper, but it quires users to handle router, routedata, routerbuild and so on. I think that will make this scenario much harder.

yishaigalatzer replied:
Probably not  the best idea to subclass the `UrlHelper` subclassing is useless I hear.

rynowak commented on src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelperCallback.cs:
```diff
@@ -0,0 +1,29 @@
+		{
+			get;
+			set;
+		}
```
Take a look at https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNet.Http.Core/DefaultHttpContext.cs

In general all of the `HttpContext` stuff is get/set properties. In general setting a property is much simpler to get your head around than using callbacks, so we should leverage that there the default is testable. If you find things in that space that don't work well let's loop in the appropriate folks and improve it.

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs:
```diff
@@ -0,0 +1,277 @@
+{
+	public class MvcControllerUnitTestHelper
+	{
+		public MvcControllerUnitTestHelper() : this(new ServiceCollection(), new RouteData())
```
Completely pointless, but the spacing here looks odd.

yishaigalatzer commented on src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs:
```diff
@@ -0,0 +1,277 @@
+			set;
+		}
+
+		public Controller Initialize(Controller controller, MvcControllerUnitTestHelperCallback callback)
```
a generic method will be much more useful

ianhong replied:
@yishaigalatzer I think non-generic version can be more useful because many of controllers take parameters for IDbContxt, or ICacheManager and so on..
Do you think we should provide both of non-generic version and generic version?

yishaigalatzer commented on test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs:
```diff
@@ -0,0 +1,173 @@
+using Microsoft.AspNet.Http.Core.Collections;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.Core.Test
```
This will move to a new test project as well

yishaigalatzer commented on test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs:
```diff
@@ -0,0 +1,173 @@
+			};
+
+			// Act
+			new MvcControllerUnitTestHelper().Initialize(controller, callback);
```
this will just be:

```c#
ControllerTestHelper.Create<TestabilityController)(callback);
```

or - 

```c#
ControllerTestHelper.Create<TestabilityController)(new ControllerTestCallback() // note the name change
    {
        On RequestAborted = () => new CancellationToken(true),
    });
```

yishaigalatzer commented on test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs:
```diff
@@ -0,0 +1,173 @@
+			var viewResult = controller.UsingContextRequestAborted();
+
+			// Assert
+			Assert.Equal("RequestAborted", viewResult.ViewName);
```
This is a bad example - if the Request is aborted, you will never return a view

ianhong commented:
@rynowak and @yishaigalatzer 

I updated this PR per your comments. This is the summary:
1) Changed the IUrlHeper interface to remove IsLocalUrl and made it a static method inside UrlHelper. I think I should send a seperate PR for this. 
2) The --Callback name changed to --Context to accommodate the simple value properties.
3) Used DefaultHttpContext 
4) Helper is now a static class

ianhong commented on src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs:
```diff
@@ -0,0 +1,98 @@
+				{
+					helperContext.RequestScheme = protocol;
+					helperContext.RequestHost = host;
+					helperContext.RequestFragment = fragment;
```
I updated helperContext here to simplify the OnUrlAction callback. not sure if this a good design or not. Otherwise, the signature of OnUrlAction will be Func<string, string, object, string, string, string, string> OnUrlAction. I think this is too long for users.

rynowak replied:
My suggestion is to make `MockUrlHelper` a property on the `HelperContext` that gets initialized by default. Then you can add multiple overloads of things like `OnAction` to that class.

ianhong replied:
@rynowak How about converting OnUrlAction property to the methods of HelperContext such as. In this way, we don't have to make MockUrlHelper public...?? 
class HelperContext {
               // 
		public void OnUrlAction(Func<string> actionCallBack)
		{
			((MockUrlHelper)UrlHelper).ActionDelegate = actionCallBack;
		}

		public void OnUrlAction(Func<string, string> actionCallBack)
		{
			((MockUrlHelper)UrlHelper).ActionDelegate = actionCallBack;
		}
		public void OnUrlAction(Func<string, string, object> actionCallBack)
		{
			((MockUrlHelper)UrlHelper).ActionDelegate = actionCallBack;
		}
}

rynowak replied:
Both of these would certainly work. I would also add that `don't have to make MockUrlHelper public` isn't a design goal, if there's a benefit to making it public then we'll go that route.

Sometimes context is good.

`context.Url.OnAction(...)` vs `context.OnUrlAction(...)`

ianhong commented on test/Microsoft.AspNet.Mvc.Testing.Test/ControllerTestHelperTest.cs:
```diff
@@ -0,0 +1,198 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
```
In this design PR, I didn't make these test more realistic. I'll do it in the actual PR.

rynowak replied:
It's **really** hard to get a grasp on whether or not this design stands up without some compelling tests and controller code to look at.

Examples:
https://github.com/aspnet/MusicStore/blob/master/src/MusicStore/Controllers/AccountController.cs#L227
https://github.com/aspnet/MusicStore/blob/master/src/MusicStore.Spa/Apis/AlbumsApiController.cs#L144

ianhong replied:
@rynowak 2 examples that you pointed out are really full-blown controllers. For now, I'm going to use Mock as much as possible for the all external components such as EF, UserManager, and so on. Plus, those tests are not going to be checked in the repository, but they are just to show how users codes look like..still I think it will takes some time..

ianhong replied:
@rynowak I looked into two controllers, AccountController and AlubmsApiController. I don't think we are interested in all the actions of the controllers. Many of them doesn't even need this feature to test. 
 And, I think the following actions should be good enough to show how users codes look like:
1. VerifyCode for using ModelState.isValid and ModelState.AddModelError
2. Register for Url.Action, RedirectToAction, ViewBag.Link View, AddErrors
3. ExternalLoginConfirmation for User.Identity.IsAuthenticated and  ModelState.IsValid

Let me know if you have any other actions to try out.

ianhong commented on test/Microsoft.AspNet.Mvc.Testing.Test/ControllerTestHelperTest.cs:
```diff
@@ -0,0 +1,198 @@
+			var viewResult = controller.UsingContextRequestAborted();
+
+			// Assert
+			Assert.Null(viewResult);
```
What is the correct behavior when requests are aborted?? This Context.RequestAborted is used in many places in MusicStore like https://github.com/aspnet/MusicStore/blob/master/src/MusicStore/Controllers/CheckoutController.cs
I can't find any Http status code for it too.

rynowak commented on src/Microsoft.AspNet.Mvc.Core/UrlHelper.cs:
```diff
@@ -68,8 +68,13 @@ public UrlHelper(IScopedInstance<ActionContext> contextAccessor, IActionSelector
+        /// Returns a value that indicates whether the URL is local.
+        /// </summary>
+        /// <param name="url">The URL.</param>
+        /// <returns>true if the URL is local; otherwise, false.</returns>
```
`<c>true</c>` same for false.

rynowak commented on src/Microsoft.AspNet.Mvc.Core/UrlHelper.cs:
```diff
@@ -68,8 +68,13 @@ public UrlHelper(IScopedInstance<ActionContext> contextAccessor, IActionSelector
+        /// <param name="url">The URL.</param>
+        /// <returns>true if the URL is local; otherwise, false.</returns>
+
+        public static bool IsLocalUrl(string url)
```
@yishaigalatzer we were discussing making this method back into a static. It's a security feature and having it show up as mockable seemed bad.

rynowak commented on src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs:
```diff
@@ -0,0 +1,98 @@
+		public static Controller Initialize(Controller controller, ControllerTestHelperContext context)
+		{
+			return Initialize(controller, new ActionDescriptor(), context);
+			// Question: Can we build (fully or partially) ActionDescriptor() from tyep information??
```
To answer your question, we can't do this in any sort of accurate way.

rynowak commented on src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs:
```diff
@@ -0,0 +1,98 @@
+			featureCollection.Add(typeof(IHttpRequestLifetimeFeature), new FeatureContext());
+			featureCollection.Add(typeof(IHttpRequestFeature), new FeatureContext());
+
+			var httpContext = new DefaultHttpContext(featureCollection)
```
None of this should be needed. Just `new DefaultHttpContext()`

ianhong replied:
The main reason of using this feature collection is providing users with a way to mock  Context.RequestAborted.IsCancellationRequested. I'll talk to @Tratcher if there is a better way to do this. In the meantime I'd like to keep this to unblock myself..

Tratcher replied:
```C#
var context = new DefaultHttpContext();
context.SetFeature<IHttpRequestLifetimeFeature>(myMockLifetimeFeature);
...
class MyMockLifetimeFeature : IHttpRequestLifetimeFeature
{
...
}
```

ianhong replied:
@Tratcher Thanks,
That looks not much different from new DefaultHttpContext(myMockLifetimeFeature)..I think what @rynowak meant was is there any simpler way to do that like RequestAborted settable properties?

Tratcher replied:
If you try my sample it will be simpler than what you already have, you don't have to mock IHttpRequestFeature.

Making HttpContext.RequestAborted and/or IHttpRequestLifetime.RequestAborted settable could be done, but we'd have to work out what that would mean for IHttpRequestLifetimeFeature.Abort().

rynowak commented on src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs:
```diff
@@ -0,0 +1,98 @@
+		}
+
+		private static Controller Initialize(Controller controller, ActionDescriptor actionDescriptor,
+			ControllerTestHelperContext helperContext)
```
don't break the line in the middle of the parameter list. Either they all go on their own line, or all on the same.

rynowak commented on src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs:
```diff
@@ -0,0 +1,98 @@
+			var httpContext = new DefaultHttpContext(featureCollection)
+			{
+				RequestServices = serviceProvider
+			};
```
We should probably set `ApplicationServices` also

rynowak commented on src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelperContext.cs:
```diff
@@ -0,0 +1,76 @@
+		{
+			get;
+			set;
+		}
```
These are both absolutely useful. My suggestion for all of these is to pre-intialize them so the user can just set values directly.

rynowak commented on src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelperContext.cs:
```diff
@@ -0,0 +1,76 @@
+		{
+			get;
+			set;
+		}
```
Both of these are potentially useful - `ServiceCollection` to manually add services that will be resolved.

ianhong replied:
@rynowak I think this design has a problem when users call different overloaded Action methods in a controller. 1) It would make us to maintain a list of delegates for overloaded methods, 2) it makes user to write different callbacks for different overloaded delegates. 
Instead, I'd like to create a context class like
public class UrlHelperContext
{
    string Action {get;set;} string Controller {get; set;...}
}
And ..public Func<UrlHelperContext, string> OnUrlAction { get; set; }

IUrlHeper.RouteUrl method requires the same parameters, I think it makes APIs simpler after all. 

Let me know if you have any concerns on this.

rynowak replied:
Would love to see you try this out and show an example. Give it a shot

rynowak commented on src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelperContext.cs:
```diff
@@ -0,0 +1,76 @@
+		{
+			get;
+			set;
+		}
```
If you pre-intialize `HttpContext`, then there's no real need for any of these, because you can just do it through the context.

`context.HttpContext.Request.Scheme = "https";`

rynowak commented on src/Microsoft.AspNet.Mvc.Testing/FeatureContext.cs:
```diff
@@ -0,0 +1,179 @@
+
+namespace Microsoft.AspNet.Mvc.Testing
+{
+	internal class FeatureContext :
```
this shouldn't be needed

rynowak commented on src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelperContext.cs:
```diff
@@ -0,0 +1,76 @@
+		{
+			get;
+			set;
+		}
```
suggest removing this and adding `MockUrlHelper` directly.

rynowak commented on test/Microsoft.AspNet.Mvc.Testing.Test/ControllerTestHelperTest.cs:
```diff
@@ -0,0 +1,198 @@
+					Assert.Equal("fragment1", helperContext.RequestFragment);
+					Assert.Equal("redirect", action);
+					return action;
+				}
```
This isn't really right - why is the helperContext being validated in a test? You want to validate that route values passed to URL helper are valid, and then generate a fake URL so the action can proceed.

ianhong commented:
@yishaigalatzer and @rynowak 

I think this Design PR has all the feature to support MusicStore and KEpense except:

1. Context.User.Identity.GetUserId(),
2. Content.Request.Cookies

1. Contxt.User.Identity.GetUserId will be easy to support with this design
2. I haven't looked into Cookies yet. 

Thanks,

ianhong commented:
Opened #1935 and #1934 for more specific work items, and closing this one.
